src/gitingest/utils/s3_utils.py

```
@@ -0,0 +1,212 @@
+    """Get S3 bucket name from environment variables."""
+    return os.getenv("S3_BUCKET_NAME", "gitingest-bucket")
+
+
+def get_s3_alias_host() -> str | None:
+    """Get S3 alias host for public URLs."""
+    return os.getenv("S3_ALIAS_HOST")
+
+
+def generate_s3_file_path(
+    source: str,
+    user_name: str,
+    repo_name: str,
+    branch: str | None,
+    commit: str | None,
+    include_patterns: set[str] | None,
+    ignore_patterns: set[str],
+) -> str:
+    """Generate S3 file path with proper naming convention.
+
+    Format: /ingest/<provider>/<repo-owner>/<repo-name>/<branch>/<commit-ID>/<exclude&include hash>.txt
+    The commit-ID is always included in the URL. If no specific commit is provided,
+    the actual commit hash from the cloned repository is used.
+
+    Args:
+        source: Git host (github, gitlab, etc.)
+        user_name: Repository owner/user
+        repo_name: Repository name
+        branch: Branch name (if available)
+        commit: Commit hash (should always be available now)
+        include_patterns: Include patterns set
+        ignore_patterns: Ignore patterns set
+
+    Returns:
+        S3 file path string
+
+    """
+    # Extract source from URL or default to "unknown"
+    if "github.com" in source:
+        git_source = "github"
+    elif "gitlab.com" in source:
+        git_source = "gitlab"
+    elif "bitbucket.org" in source:
+        git_source = "bitbucket"
+    else:
+        git_source = "unknown"
+
+    # Use branch, fallback to "main" if neither branch nor commit
+    branch_name = branch or "main"
+
+    # Create hash of exclude/include patterns for uniqueness
+    patterns_str = f"include:{sorted(include_patterns) if include_patterns else []}"
+    patterns_str += f"exclude:{sorted(ignore_patterns)}"
+
+    patterns_hash = hashlib.sha256(patterns_str.encode()).hexdigest()[:16]
+
+    # Commit should always be available now, but provide fallback just in case
+    commit_id = commit or "HEAD"
```

We `commit` should always be available, I think we should raise an error if not.